### PR TITLE
Closed Loop on Wrist

### PR DESCRIPTION
**Pull request by NickVerwey**

### CODE REVIEW

File: src/main/java/frc/robot/Robot.java [hunk 1]
```diff
@@ -16,7 +19,13 @@
 
   public static Drivetrain mDrivetrain = Drivetrain.create();
   public static Elevator mElevator = Elevator.create();
+  public static Wrist mWrist = Wrist.create();
+  // public static LEDs mLeds = LEDs.create();
+  // public static Cargo mCargo = Cargo.create();
```
Comment (ghost) [inline src/main/java/frc/robot/Robot.java]: Can this be deleted?

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -16,7 +19,13 @@
...
+  // public static Panel mPanel = Panel.create();
```
Comment (ghost) [inline src/main/java/frc/robot/Robot.java]: Can this also be deleted?

File: src/main/java/frc/robot/commands/cargo/CargoIntake.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
+package frc.robot.commands.cargo;
+
+import edu.wpi.first.wpilibj.command.Command;
+import frc.robot.Robot;
+
+public class CargoIntake extends Command {
```
Comment (ghost) [inline src/main/java/frc/robot/commands/cargo/CargoIntake.java:6]: Could this be an instant command?

File: src/main/java/frc/robot/commands/wrist/WristJoystick.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  /****************************************************************************************************************************** 
+  ** COMMAND OVERRIDES
+  ******************************************************************************************************************************/
+  @Override
+  protected void execute() {
+    // Note that on an Xbox Controller, a positive input is to the bottom-right
+    double zWrist = -(OI.mOperatorJoystick.getY() * 0.5);
+
+    // Apply a deadband to the joystick
+    if (zWrist < 0.05 && zWrist > -0.05) {
```
Comment (ghost) [inline src/main/java/frc/robot/commands/wrist/WristJoystick.java]: Might want to add the deadband to the robotmap

File: src/main/java/frc/robot/commands/elevator/ElevatorJoystick.java [hunk 5]
```diff
@@ -22,7 +28,8 @@ public ElevatorJoystick() {
   ******************************************************************************************************************************/
   @Override
   protected void execute() {
-    double zElevator = OI.mOperatorJoystick.getY();
+    // Note that on an Xbox Controller, a positive input is to the bottom-right
+    double zElevator = -OI.mOperatorJoystick.getRawAxis(5);
 
     // Apply a deadband to the joystick
     if (zElevator < 0.05 && zElevator > -0.05) {
```
Comment (ghost) [inline src/main/java/frc/robot/commands/elevator/ElevatorJoystick.java]: Might want to add the deadband to the robotmap

File: src/main/java/frc/robot/commands/wrist/WristJoystick.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    requires(Robot.mWrist);
+  }
+
+  /****************************************************************************************************************************** 
+  ** COMMAND OVERRIDES
+  ******************************************************************************************************************************/
+  @Override
+  protected void execute() {
+    // Note that on an Xbox Controller, a positive input is to the bottom-right
+    double zWrist = -(OI.mOperatorJoystick.getY() * 0.5);
```
Comment (ghost) [inline src/main/java/frc/robot/commands/wrist/WristJoystick.java]: Might want to add that gain to the robotmap

File: src/main/java/frc/robot/commands/wrist/WristToAngle.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+
+  /****************************************************************************************************************************** 
+  ** COMMAND OVERRIDES
+  ******************************************************************************************************************************/
+  @Override
+  protected void initialize() {
+    // mLogger.info("Starting WristToAngle command: [{}]", mTargetAngle);
+    // // setInterruptible(true);
+    // Robot.mWrist.MotionMagicOutput(Robot.mWrist.degreesToSensorTicks(mTargetAngle) * RobotMap.kWristGearing + 100);
+    Robot.mWrist.MotionMagicOutput(3300 + RobotMap.kWristTickGain);
```
Comment (ghost) [inline src/main/java/frc/robot/commands/wrist/WristToAngle.java:34]: Looks like we still need to generate an equation

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 8]
```diff
@@ -165,26 +171,31 @@ public void ApplyDriveSignal(double throttle, double turn) {
...
-      if (turn > 0.0) {
-        mTurn = RobotMap.kDeadbandLowGear + kDeadbandLowGearScalar * turn;
-        mThrottle = RobotMap.kDeadbandLowGear + kDeadbandLowGearScalar * throttle;
-      } else {
-        mTurn = -1.0 * RobotMap.kDeadbandLowGear + kDeadbandLowGearScalar * turn;
-        mThrottle = -1.0 * RobotMap.kDeadbandLowGear + kDeadbandLowGearScalar * throttle;
-      }
-    } else {
-      mLogger.error("Unknown state of shifter [{}]", mIsHighGear);
+    // if (mIsHighGear) {
```
Comment (ghost) [inline src/main/java/frc/robot/subsystems/Drivetrain.java:174]: We will revisit this once we get up to Duluth,

File: src/main/java/frc/robot/subsystems/MultiManipulator.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+  public MultiManipulator(WPI_TalonSRX cargoMotor, DoubleSolenoid panelActuator) {
+    mCargoMotor = cargoMotor;
+    mCargoMotor.setInverted(true);
+    mCargoMotor.setNeutralMode(NeutralMode.Coast);
+
+    mPanelActuator = panelActuator;
+    shiftPanelIntake(true);
+
+    // Current limiting
+    mCargoMotor.configContinuousCurrentLimit(10, RobotMap.kLongCANTimeoutMs);
```
Comment (ghost) [inline src/main/java/frc/robot/subsystems/MultiManipulator.java:76]: Thanks for remembering to add this.  I think we can limit the current even more.

Comment (ghost) [reply src/main/java/frc/robot/subsystems/MultiManipulator.java:76]: https://motors.vex.com/vexpro-motors/775pro

File: src/main/java/frc/robot/subsystems/MultiManipulator.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+    mCargoMotor.configPeakCurrentLimit(40, RobotMap.kLongCANTimeoutMs);
+    mCargoMotor.configPeakCurrentDuration(200, RobotMap.kLongCANTimeoutMs);
+    mCargoMotor.enableCurrentLimit(true);
+
+    mLogger.info("Cargo Manipulator Created");
+  }
+
+  public static MultiManipulator create() {
```
Comment (ghost) [inline src/main/java/frc/robot/subsystems/MultiManipulator.java:84]: You've got all of these constructors setup for injecting.  Time to start learning Junit and Mockito :)

File: src/main/java/frc/robot/subsystems/Wrist.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+  ** CONSTRUCTOR
+  ******************************************************************************************************************************/
+  public Wrist(WPI_TalonSRX master) {
+    mMaster = master;
+
+    //ErrorCode errorCode;
+
+
+    // Configure the feedback sensor
+    mMaster.configSelectedFeedbackSensor(FeedbackDevice.CTRE_MagEncoder_Relative, RobotMap.kPIDLoopIdx, RobotMap.kLongCANTimeoutMs);
```
Comment (ghost) [inline src/main/java/frc/robot/subsystems/Wrist.java:114]: For all of these Talon setup commands we should be checking their return status

**Review by ghost — COMMENTED** [review]
Just a few minor things to fix before checking in.

File: src/main/java/frc/robot/commands/wrist/WristToAngle.java (hunk 7 above)
Comment (NickVerwey) [reply src/main/java/frc/robot/commands/wrist/WristToAngle.java:34]: Line 33 is the theoretic equation, but it produces a much less stable wrist than the manual declaration

File: src/main/java/frc/robot/commands/cargo/CargoIntake.java (hunk 3 above)
Comment (NickVerwey) [reply src/main/java/frc/robot/commands/cargo/CargoIntake.java:6]: We could change it to whileheld as an instant command. Would it still need a whenreleased if we did that?

Comment (NickVerwey) [reply src/main/java/frc/robot/commands/cargo/CargoIntake.java:6]: Same sentiment applies to all cargo commands

File: src/main/java/frc/robot/subsystems/Wrist.java (hunk 11 above)
Comment (NickVerwey) [reply src/main/java/frc/robot/subsystems/Wrist.java:114]: Doing the ErrorCode checking, you mean? As in, 
if (ErrorCode != ErrorCode.ok) {
  mLogger.info("Error");
}

File: src/main/java/frc/robot/commands/cargo/CargoIntake.java (hunk 3 above)
Comment (ghost) [reply src/main/java/frc/robot/commands/cargo/CargoIntake.java:6]: you mean a whenreleased to stop the motor, right?

File: src/main/java/frc/robot/commands/wrist/WristToAngle.java (hunk 7 above)
Comment (ghost) [reply src/main/java/frc/robot/commands/wrist/WristToAngle.java:34]: If (3300 + RobotMap.kWristTickGain) is the same as (Robot.mWrist.degreesToSensorTicks(mTargetAngle) * RobotMap.kWristGearing + 100) for holding the wrist horizontal, then there's no reason the stability will be different
Did you verify these were the same?

File: src/main/java/frc/robot/subsystems/Wrist.java (hunk 11 above)
Comment (ghost) [reply src/main/java/frc/robot/subsystems/Wrist.java:114]: yes

**Comment by ghost** [comment]
I'm concerned about the wrist.  I was hoping to add in the other commands for the different angles.

**Review by ghost — APPROVED** [review]
Go ahead and merge.  We will still be making changes once we get to Duluth.

**Pull request merged**